cpp/src/arrow/builder.cc

```diff
@@ -121,6 +121,14 @@ void ArrayBuilder::UnsafeAppendToBitmap(const uint8_t* valid_bytes, int64_t leng
   uint8_t bitset = null_bitmap_data_[byte_offset];
 
   for (int64_t i = 0; i < length; ++i) {
+    if (bit_offset == 8) {
```

Comment:
Might want to look at other places where this fast-bitsetting logic exists to see if there might be an inappropriate memory access